### ISSUE TITLE
Batch Gradient Decent

### ISSUE THREAD

**Issue by cmcgrath1982** [ageron/handson-ml#270]
This may not be the best place to ask this, but I couldn't find another place to ask this. 

I'm working through Chapter 4 on the batch gradient decent sections. How does the code in the book know how to stop at the correct solution for any of the methods? It iterates through 1000 repetitions calculating theta as it goes, but what eventually stops it at the solution? I'm looking at the code at the bottom of page 117 for reference. Is theta just the cumulative difference?

**Comment by ageron** [comment]
Hi @cmcgrath1982 ,

Thanks for your question.  The short answer is that we don't know whether we have stopped too early or too late.  If you stop too early, then the algorithm will not have had time to converge to a good solution.  If you stop too late, then there is a risk that it will have overfitted the training set.  That said, since the models are pretty simple linear models, it is highly unlikely that they will overfit the training set, so if you just measure the training error and see that it stops going down, it probably means that you can stop the algorithm.  However, a better way to know when to stop is to compute the *validation* error at each epoch and stop training when it reaches a minimum. This is called "early stopping", it is explained a bit later in chapter 4.
Hope this helps,
Aurélien

**Comment by cmcgrath1982** [comment]
Thanks Aurelien, appreciate the help
